### ISSUE TITLE
[Tests] Enable training loop tests with SimpleMLP and ExTensor.randn

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2728]
## Summary
Enable comprehensive training loop tests by implementing dependencies.

## Components Needed
1. \`SimpleMLP\` - simple multi-layer perceptron model (Model trait implementation)
2. \`ExTensor.randn(shape, dtype, seed)\` - random normal initialization
3. \`create_simple_model()\` - test fixture
4. \`create_mock_dataloader(n_batches)\` - test fixture
5. \`create_simple_dataset()\` - test fixture

## Files Affected
- \`tests/shared/training/test_training_loop.mojo\` (lines 94-123)

## Related TODO Lines in todo.log
- Lines 94-123: 30 TODOs in test_training_loop.mojo
- All 22 test functions currently stubbed with \`pass\`

## Notes
All training loop tests are blocked on these components.

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have a complete understanding of the codebase. Let me create the detailed implementation plan:

---

# Implementation Plan: Enable Training Loop Tests with SimpleMLP and ExTensor.randn

## Issue #2728

---

## 1. Summary

This plan addresses enabling 13 comprehensive training loop tests that are currently stubbed with `pass` statements. The tests are well-documented and define the expected API contracts for the training infrastructure.

### Key Findings from Analysis

| Component | Current Status | Action Required |
|-----------|---------------|-----------------|
| `ExTensor.randn()` | ✅ **Already Implemented** | Export to tests via `shared.core` |
| `SimpleMLP` | ✅ **Fully Implemented** | Already implements `Model` trait |
| `create_simple_model()` | ✅ **Fully Implemented** | Returns `SimpleMLP(10, 5, 1)` |
| `create_mock_dataloader()` | ✅ **Fully Implemented** | Returns `MockDataLoader` |
| `create_simple_dataset()` | ✅ **Fully Implemented** | Returns `List[Tuple[...]]` |
| `SGD.step()` | ⚠️ **Stub Only** | Implement parameter updates |
| `MSELoss.compute()` | ⚠️ **Stub Only** | Implement actual MSE computation |
| Gradient System (`_grad`) | ❌ **Not Implemented** | Out of scope (separate issue #2725) |

**Critical Discovery**: The `randn` function already exists at `/home/mvillmow/ml-odyssey/shared/core/extensor.mojo:2765` with Box-Muller transform, seed support, and full dtype support. The tests just need updated imports.

---

## 2. Step-by-Step Implementation Tasks

### Phase 1: Export `randn` to Tests (Priority: High)

**Task 1.1**: Add `randn` export to `shared/core/__init__.mojo`

- **Why**: Tests import from `shared.core` but `randn` is defined in `extensor.mojo`
- **Action**: Add `randn` to the export list in `shared/core/__init__.mojo`
- **Verification**: `from shared.core import randn` compiles successfully

**Task 1.2**: Update test imports in `test_training_loop.mojo`

- **Why**: Tests need access to `randn` function
- **Action**: Add `randn` to imports from `shared.core`
- **Current**: `from shared.core import ones, zeros`
- **Updated**: `from shared.core import ones, zeros, randn`

---

### Phase 2: Implement `MSELoss.compute()` (Priority: High)

**Task 2.1**: Implement actual MSE computation in `shared/training/__init__.mojo`

- **Location**: Lines 133-157
- **Current**: Returns dummy `zeros(List[Int](), DType.float32)`
- **Required**: `MSE = mean((pred - target)^2)`

```mojo
fn compute(self, pred: ExTensor, target: ExTensor) raises -> ExTensor:
    """Compute MSE loss between predictions and targets."""
    # Element-wise squared difference
    var diff = pred - target  # Requires ExTensor.__sub__
    var squared = diff * diff  # Requires ExTensor.__mul__
    
    # Mean reduction
    var total = Float64(0.0)
    for i in range(squared.numel()):
        total += squared._get_float64(i)
    var mean_val = total / Float64(squared.numel())
    
    # Return scalar tensor
    var result = zeros(List[Int](), DType.float32)
    result._set_float32(0, Float32(mean_val))
    return result^
```

**Verification**: `test_training_loop_computes_loss()` passes with expected MSE ≈ 4.67

---

### Phase 3: Implement `SGD.step()` (Priority: High)

**Task 3.1**: Implement parameter update logic in `shared/training/__init__.mojo`

- **Location**: Lines 85-101
- **Current**: `pass` (no-op)
- **Required**: `param = param - learning_rate * grad`

**Note**: Full gradient-based updates require the gradient system (#2725). For this issue, implement basic parameter update infrastructure:

```mojo
fn step(mut self, params: List[ExTensor]) raises:
    """Update parameters using gradients.
    
    Note: Actual gradient-based updates blocked on #2725.
    This implementation provides the infrastructure.
    """
    # Placeholder until gradient system ready
    # Once _grad attribute exists on ExTensor:
    # for i in range(len(params)):
    #     var param = params[i]
    #     var grad = param._grad
    #     for j in range(param.numel()):
    #         var new_val = param._get_float32(j) - self.learning_rate * grad._get_float32(j)
    #         param._set_float32(j, new_val)
    pass
```

---

### Phase 4: Update Test File to Uncomment Tests (Priority: High)

**Task 4.1**: Enable tests that don't require gradient system

Tests that can be enabled immediately (no gradient dependency):

| Test Function | Lines | Dependencies Met |
|--------------|-------|------------------|
| `test_training_loop_single_batch()` | 51-88 | ✅ SimpleMLP, MSELoss, SGD |
| `test_training_loop_full_epoch()` | 91-115 | ⚠️ Needs MockDataLoader→ExTensor |
| `test_training_loop_forward_pass()` | 152-173 | ✅ randn, SimpleMLP |
| `test_training_loop_forward_batches_independently()` | 176-198 | ✅ randn, SimpleMLP |
| `test_training_loop_computes_loss()` | 206-247 | ✅ MSELoss.compute() |
| `test_training_loop_loss_scalar()` | 250-268 | ✅ randn, MSELoss |
| `test_training_loop_processes_variable_batch_sizes()` | 426-448 | ✅ randn, SimpleMLP |
| `test_training_loop_averages_loss_over_batch()` | 451-486 | ✅ randn, MSELoss |

Tests that require gradient system (blocked on #2725):

| Test Function | Lines | Blocking Issue |
|--------------|-------|----------------|
| `test_training_loop_backward_pass()` | 276-306 | Needs `_grad` attribute |
| `test_training_loop_gradient_accumulation()` | 309-341 | Needs `_grad` attribute |
| `test_training_loop_updates_weights()` | 349-379 | Needs gradient system |
| `test_training_loop_respects_learning_rate()` | 382-418 | Needs gradient system |
| `test_training_loop_property_loss_decreases_on_simple_problem()` | 494-521 | Needs full training loop |
| `test_training_loop_multiple_epochs()` | 118-144 | Needs gradient system |

**Task 4.2**: Update test file with proper imports and uncommented code

Replace `TODO(#2728)` comments with actual test code for tests that can be enabled.

---

### Phase 5: Add MockDataLoader → ExTensor Conversion (Priority: Medium)

**Task 5.1**: Add batch extraction method to `MockDataLoader`

The `run_epoch()` method needs ExTensor batches, but `MockDataLoader` stores `List[Float32]`.

```mojo
# In tests/shared/conftest.mojo
fn get_batch(self, batch_idx: Int) raises -> Tuple[ExTensor, ExTensor]:
    """Get batch as ExTensor pair.
    
    Args:
        batch_idx: Index of batch to retrieve.
    
    Returns:
        Tuple of (inputs, targets) as ExTensors.
    """
    var start = batch_idx * self.batch_size
    var end = min(start + self.batch_size, len(self.samples))
    var batch_size = end - start
    
    # Create input tensor
    var input_dim = len(self.samples[0][0])
    var inputs = zeros([batch_size, input_dim], DType.float32)
    
    # Create target tensor
    var output_dim = len(self.samples[0][1])
    var targets = zeros([batch_size, output_dim], DType.float32)
    
    # Fill tensors
    for i in range(batch_size):
        var sample = self.samples[start + i]
        for j in range(input_dim):
            inputs._set_float32(i * input_dim + j, sample[0][j])
        for j in range(output_dim):
            targets._set_float32(i * output_dim + j, sample[1][j])
    
    return (inputs^, targets^)
```

---

## 3. Files to Modify/Create

### Files to Modify

| File | Changes | Lines |
|------|---------|-------|
| `shared/core/__init__.mojo` | Add `randn` to exports | ~5 lines |
| `shared/training/__init__.mojo` | Implement `MSELoss.compute()` | ~15 lines |
| `shared/training/__init__.mojo` | Add TODO note for `SGD.step()` | ~5 lines |
| `tests/shared/training/test_training_loop.mojo` | Uncomment enabled tests, update imports | ~200 lines |
| `tests/shared/conftest.mojo` | Add `get_batch()` to MockDataLoader | ~30 lines |

### Files NOT to Modify (Out of Scope)

| File | Reason |
|------|--------|
| `shared/core/extensor.mojo` | `randn` already implemented |
| `shared/testing/models.mojo` | `SimpleMLP` already complete |
| `shared/core/traits.mojo` | Model/Loss/Optimizer traits complete |

---

## 4. Testing Approach

### Test Execution Order

1. **Unit Tests**: Run individual test functions to verify components
2. **Integration Tests**: Run full test file with `mojo test`
3. **CI Verification**: Ensure tests pass in CI pipeline

### Verification Commands

```bash
# Verify randn export
cd /home/mvillmow/ml-odyssey && pixi run mojo build -c "from shared.core import randn"

# Run specific test
cd /home/mvillmow/ml-odyssey && pixi run mojo test tests/shared/training/test_training_loop.mojo

# Run all training tests
cd /home/mvillmow/ml-odyssey && pixi run mojo test tests/shared/training/
```

### Expected Test Results

| Test Category | Tests Enabled | Tests Blocked | Notes |
|--------------|---------------|---------------|-------|
| Core Training | 1 of 3 | 2 | Blocked on gradient system |
| Forward Pass | 2 of 2 | 0 | Fully enabled |
| Loss Computation | 2 of 2 | 0 | Fully enabled |
| Backward Pass | 0 of 2 | 2 | Blocked on gradient system |
| Weight Update | 0 of 2 | 2 | Blocked on gradient system |
| Batch Processing | 2 of 2 | 0 | Fully enabled |
| Property-Based | 0 of 1 | 1 | Blocked on gradient system |
| **Total** | **7 of 13** | **6** | 54% enabled |

---

## 5. Success Criteria

### Minimum Success (Issue Complete)

- [ ] `randn` exported from `shared.core`
- [ ] `MSELoss.compute()` returns correct MSE values
- [ ] At least 7 tests uncommented and passing
- [ ] All uncommented tests pass in CI
- [ ] Blocked tests have clear TODO references to #2725

### Full Success (Stretch Goal)

- [ ] All non-gradient tests enabled (7 tests)
- [ ] `MockDataLoader.get_batch()` implemented
- [ ] Test file compiles with no warnings
- [ ] All fixture functions verified working

### Verification Checklist

```markdown
- [ ] `from shared.core import randn` compiles
- [ ] `test_training_loop_computes_loss()` passes with MSE ≈ 4.67
- [ ] `test_training_loop_forward_pass()` verifies batch dimension preservation
- [ ] `test_training_loop_loss_scalar()` verifies Float32 return type
- [ ] All enabled tests pass: `pixi run mojo test tests/shared/training/test_training_loop.mojo`
- [ ] CI pipeline passes
```

---

## 6. Dependencies and Risks

### Dependencies

| Dependency | Status | Impact |
|------------|--------|--------|
| ExTensor arithmetic (`-`, `*`) | ✅ Implemented | Required for MSELoss |
| ExTensor.randn() | ✅ Implemented | Required for tests |
| SimpleMLP Model trait | ✅ Implemented | Required for tests |
| Gradient system (#2725) | ❌ Not ready | Blocks 6 tests |

### Risks and Mitigations

| Risk | Likelihood | Impact | Mitigation |
|------|------------|--------|------------|
| ExTensor arithmetic not working | Low | High | Verify with simple test first |
| Import issues with randn | Medium | Low | Check __init__.mojo exports |
| MSELoss precision issues | Low | Medium | Use tolerance in assertions |

---

## 7. Implementation Sequence

```
Week 1:
├─ Day 1: Export randn, update test imports
├─ Day 2: Implement MSELoss.compute()
├─ Day 3: Uncomment 7 enabled tests
├─ Day 4: Add MockDataLoader.get_batch()
└─ Day 5: Verify CI, create PR

Follow-up (separate issue):
└─ Implement gradient system (#2725) to enable remaining 6 tests
```

---

## Resource Usage

- **Tool calls used**: 14
- **Steps taken**: 6
- **Budget remaining**: 36 tool calls, 44 steps

**Comment by mvillmow** [comment]
## Dependencies (AI-Analyzed)

**Priority**: P1

### Depends On
- #2717
- #2719
- #2721
- #2722
- #2725
- #2726

### Analysis
Training loop tests depend on matrix ops (#2717), element-wise math (#2719), slicing (#2721), utility methods (#2722), gradient system (#2725), and checkpointing (#2726).

---
*Analyzed by Claude Code dependency analyzer*

**Comment by mvillmow** [comment]
✅ **Dependency resolved**: Issue #2665 (Learning Rate Scheduling) is now complete. All LR schedulers are implemented and tested. Training loop tests can now integrate scheduler functionality.
